programs/vote/src/vote_state/mod.rs

```diff
@@ -877,8 +877,27 @@ pub fn update_commission<S: std::hash::BuildHasher>(
     vote_account: &mut BorrowedAccount,
     commission: u8,
     signers: &HashSet<Pubkey, S>,
+    epoch_schedule: &std::sync::Arc<EpochSchedule>,
+    clock: &std::sync::Arc<Clock>,
     feature_set: &FeatureSet,
 ) -> Result<(), InstructionError> {
+    let enforce_commission_update_rule =
+        if feature_set.is_active(&feature_set::allow_commission_decrease_at_any_time::id()) {
+            is_commission_increase(vote_account, commission)?
+        } else {
+            true
+        };
+
+    #[allow(clippy::collapsible_if)]
+    if enforce_commission_update_rule
+        && feature_set
+            .is_active(&feature_set::commission_updates_only_allowed_in_first_half_of_epoch::id())
+    {
+        if !is_commission_update_allowed(clock.slot, epoch_schedule) {
+            return Err(VoteError::CommissionUpdateTooLate.into());
+        }
+    }
```

Comment:
This instruction isn't used all that often, so it's not a huge performance impact, but we should note that it's deserializing the vote account twice. It's being done once in `is_commission_increase`, and again in `update_commission`.

You did the right thing to make the change as small as possible while not introducing potential forking. We could have `is_commission_increase` take in the `VoteState` directly and update this function a bit like:

```suggestion
    let mut vote_state: VoteState = if feature_set.is_active(&feature_set::allow_commission_decrease_at_any_time::id()) {
        let mut vote_state: VoteState = vote_account
            .get_state::<VoteStateVersions>()?
            .convert_to_current();

        if is_commission_increase(&vote_state, commission)
            && feature_set
                .is_active(&feature_set::commission_updates_only_allowed_in_first_half_of_epoch::id())
        {
            if !is_commission_update_allowed(clock.slot, epoch_schedule) {
                return Err(VoteError::CommissionUpdateTooLate.into());
            }
        }
        
        vote_state
    } else {
        if feature_set
            .is_active(&feature_set::commission_updates_only_allowed_in_first_half_of_epoch::id())
        {
            if !is_commission_update_allowed(clock.slot, epoch_schedule) {
                return Err(VoteError::CommissionUpdateTooLate.into());
            }
        }
        vote_account
            .get_state::<VoteStateVersions>()?
            .convert_to_current()
    };
```

I know it's a bit more code, but it isn't ridiculously ugly. What do you think? cc @AshwinSekar for another opinion

Reply:
Yeah in the vote program we try to deserialize only once at the beginning of each instruction using either:
```rust
vote_account.get_state::<VoteStateVersions>()?.convert_to_current
```
or `verify_and_get_vote_state`

I think you could just let `is_commission_increase` take a vote state and keep it like how you have it:

```rust
    let mut vote_state: VoteState = vote_account.get_state::<VoteStateVersions>()?.convert_to_current();
    let enforce_commission_update_rule =
        if feature_set.is_active(&feature_set::allow_commission_decrease_at_any_time::id()) {
            is_commission_increase(vote_state, commission)?
        } else {
            true
        };

    #[allow(clippy::collapsible_if)]
    if enforce_commission_update_rule
        && feature_set
            .is_active(&feature_set::commission_updates_only_allowed_in_first_half_of_epoch::id())
    {
        if !is_commission_update_allowed(clock.slot, epoch_schedule) {
            return Err(VoteError::CommissionUpdateTooLate.into());
        }
    }
```

Reply:
If the vote state deserialization fails though, that could cause forking between old and new versions. old could fail during the comission check, and new could fail during the earlier deserialization, which means that two validators could give different errors for the same instruction.

Does the returned error affect consensus? I always thought that was the case

Reply:
as far as I know the returned error doesn't impact bank hash, we only check `.is_ok()`/`is_err()` in places, but I might be wrong:
https://github.com/solana-labs/solana/blob/bba6ea2d6989f8cc59ecc3a7f5ef02cb915792d9/accounts-db/src/accounts.rs#L1399-L1403

I don't mind being safe and making sure we return the same error, doesn't look too ugly.